File: 2022/solutions/day_02.py
parse = lambda x: [(ord(a)-64, ord(b)-87) for a,b in [l.split(" ") for l in x.splitlines()]]


def part_1(data):
    return sum(b + (6 if a%3+1 == b else 3 if a == b else 0) for a, b in parse(data))


def part_2(data):
    return sum(3 + a if b == 2 else 6 + a%3+1 if b == 3 else (a+1)%3+1 for a,b in parse(data))


INPUT_1 = """\
A Y
B X
C Z"""

def test():
    assert(part_1(INPUT_1) == 15)
    assert(part_2(INPUT_1) == 12)

File: 2017/src/solutions/day_02.rs
pub fn part_1(input: &str) -> String {
    input.lines().map(|l| {
        let ns = l.split_whitespace().map(|n| n.parse::<u32>().unwrap()).collect::<Vec<u32>>();
        ns.iter().max().unwrap() - ns.iter().min().unwrap()
    }).sum::<u32>().to_string()
}

pub fn part_2(input: &str) -> String {
    input.lines().map(|l| {
        let ns = l.split_whitespace().map(|n| n.parse::<u32>().unwrap()).collect::<Vec<u32>>();
        for i in 0..ns.len() {
            for j in 0..ns.len() {
                if i != j && ns[i] % ns[j] == 0 {
                    return ns[i] / ns[j];
                }
            }
        }
        0
    }).sum::<u32>().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    const EX_0: &str = "\
5 1 9 5
7 5 3
2 4 6 8";

    const EX_1: &str = "\
5 9 2 8
9 4 7 3
3 8 6 5";

    #[test]
    fn part_1_test() {
        assert_eq!(part_1(EX_0), "18");
    }

    #[test]
    fn part_2_test() {
        assert_eq!(part_2(EX_1), "9");
    }
}